File: cpp_core/tools/gyroflow_autosync.cpp
// gyroflow_autosync — UI-free C++ "autosync time" tool + timestamp-sync accuracy evaluator.
//
// Loads DJI fused-quaternion telemetry (extracted from a DJI MP4), derives an angular-velocity
// signal from the quaternions, and runs Gyroflow's time-offset finder to report the timestamp
// delay (ms) between two motion signals.
//
// Modes:
//   selftest  Inject known time offsets into a video-rate copy of the signal and measure how
//             accurately they are recovered -> timestamp-sync precision (ground-truthed).
//   compare   Find the offset between the full-rate signal and a video-rate resample of it
//             (true offset 0) -> baseline bias floor.
//   omega     Dump the quaternion-derived angular velocity as CSV (cross-check vs analysis CSV).
//
// Build: part of cpp_core (see CMakeLists.txt). Example:
//   ./gyroflow_autosync selftest --quat ../../data/dji_quaternions_full.csv --fps 30 --noise 1.5

#include "gyroflow/timesync.hpp"
#include "gyroflow/types.hpp"

#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <random>
#include <sstream>
#include <string>
#include <vector>

using namespace gyroflow;

namespace {

struct Args {
    std::string mode;
    std::string quat_path;
    double fps = 30.0;
    double search = 200.0;
    double initial = 0.0;
    double lpf = 20.0;
    bool swap_xy = false;
    double noise = 0.0;
    std::vector<double> inject = {-30.0, -15.0, -5.0, 0.0, 5.0, 15.0, 30.0};
    double range_a = 0.0;
    double range_b = 0.0; // 0,0 => whole signal
};

std::vector<double> parseList(const std::string& s) {
    std::vector<double> out;
    std::stringstream ss(s);
    std::string tok;
    while (std::getline(ss, tok, ',')) {
        if (!tok.empty()) out.push_back(std::atof(tok.c_str()));
    }
    return out;
}

void usage() {
    std::cerr <<
        "Usage: gyroflow_autosync <selftest|compare|omega> --quat <csv> [options]\n"
        "  --quat PATH      DJI quaternion CSV (dji_quaternions_full.csv or dji_camera_data.csv)\n"
        "  --fps F          video frame rate for the synthesised video-side signal (default 30)\n"
        "  --search MS      offset search size in ms (default 200)\n"
        "  --initial MS     initial/rough offset in ms (default 0)\n"
        "  --lpf HZ         low-pass cutoff in Hz (default 20)\n"
        "  --swap-xy        swap x/y when deriving angular velocity (Gyroflow estimated_gyro convention)\n"
        "  --inject LIST    selftest: comma-separated injected offsets in ms (default -30,-15,-5,0,5,15,30)\n"
        "  --noise SIGMA    selftest: Gaussian noise stddev (deg/s) added to the video signal (default 0)\n"
        "  --range A,B      analyse only timestamps in [A,B] ms (default whole signal)\n";
}

// Split a CSV line into trimmed fields.
std::vector<std::string> splitCsv(const std::string& line) {
    std::vector<std::string> out;
    std::stringstream ss(line);
    std::string f;
    while (std::getline(ss, f, ',')) {
        while (!f.empty() && (f.back() == '\r' || f.back() == ' ')) f.pop_back();
        out.push_back(f);
    }
    return out;
}

int indexOf(const std::vector<std::string>& header, const std::string& name) {
    for (std::size_t i = 0; i < header.size(); ++i) if (header[i] == name) return static_cast<int>(i);
    return -1;
}

// Auto-detect DJI CSV format and load a quaternion time series.
bool loadQuaternions(const std::string& path, std::vector<TimeQuat>& out) {
    std::ifstream in(path);
    if (!in) { std::cerr << "error: cannot open " << path << "\n"; return false; }

    std::string header_line;
    if (!std::getline(in, header_line)) { std::cerr << "error: empty file\n"; return false; }
    const std::vector<std::string> h = splitCsv(header_line);

    // Timestamp column: prefer quat_timestamp_ms (full), else timestamp_ms (camera_data).
    int ts = indexOf(h, "quat_timestamp_ms");
    if (ts < 0) ts = indexOf(h, "timestamp_ms");
    // Quaternion columns: quat_* (full) or org_quat_* (camera_data).
    int qw = indexOf(h, "quat_w"), qx = indexOf(h, "quat_x"), qy = indexOf(h, "quat_y"), qz = indexOf(h, "quat_z");
    if (qw < 0) { qw = indexOf(h, "org_quat_w"); qx = indexOf(h, "org_quat_x"); qy = indexOf(h, "org_quat_y"); qz = indexOf(h, "org_quat_z"); }

    if (ts < 0 || qw < 0 || qx < 0 || qy < 0 || qz < 0) {
        std::cerr << "error: unrecognised CSV header (need timestamp + quaternion columns)\n";
        return false;
    }

    const int need = std::max({ts, qw, qx, qy, qz});
    std::string line;
    while (std::getline(in, line)) {
        if (line.empty()) continue;
        const std::vector<std::string> f = splitCsv(line);
        if (static_cast<int>(f.size()) <= need) continue;
        TimeQuat tq;
        tq.timestamp_ms = std::atof(f[ts].c_str());
        tq.quat = Quaternion{ std::atof(f[qw].c_str()), std::atof(f[qx].c_str()),
                              std::atof(f[qy].c_str()), std::atof(f[qz].c_str()) }.normalized();
        out.push_back(tq);
    }
    // Ensure strictly increasing timestamps for interpolation/search.
    std::stable_sort(out.begin(), out.end(), [](const TimeQuat& a, const TimeQuat& b) {
        return a.timestamp_ms < b.timestamp_ms;
    });
    return !out.empty();
}

// Video-frame timestamps at `fps` spanning [a, b] ms.
std::vector<double> frameTimestamps(double a, double b, double fps) {
    std::vector<double> ts;
    if (fps <= 0.0 || b <= a) return ts;
    const double dt = 1000.0 / fps;
    for (double t = a; t <= b + 1e-6; t += dt) ts.push_back(t);
    return ts;
}

} // namespace

int main(int argc, char** argv) {
    if (argc < 2) { usage(); return 2; }
    Args a;
    a.mode = argv[1];
    for (int i = 2; i < argc; ++i) {
        std::string k = argv[i];
        auto next = [&]() -> std::string { return (i + 1 < argc) ? argv[++i] : std::string(); };
        if (k == "--quat") a.quat_path = next();
        else if (k == "--fps") a.fps = std::atof(next().c_str());
        else if (k == "--search") a.search = std::atof(next().c_str());
        else if (k == "--initial") a.initial = std::atof(next().c_str());
        else if (k == "--lpf") a.lpf = std::atof(next().c_str());
        else if (k == "--swap-xy") a.swap_xy = true;
        else if (k == "--noise") a.noise = std::atof(next().c_str());
        else if (k == "--inject") a.inject = parseList(next());
        else if (k == "--range") { auto v = parseList(next()); if (v.size() == 2) { a.range_a = v[0]; a.range_b = v[1]; } }
        else { std::cerr << "unknown arg: " << k << "\n"; usage(); return 2; }
    }
    if (a.quat_path.empty()) { std::cerr << "error: --quat is required\n"; usage(); return 2; }

    std::vector<TimeQuat> quats;
    if (!loadQuaternions(a.quat_path, quats)) return 1;

    // Full-rate angular velocity from the fused quaternions (this plays the role of the IMU gyro).
    const std::vector<TimeImu> omega = quaternionsToAngularVelocity(quats, a.swap_xy, /*degrees=*/true);

    const double first_ts = omega.front().timestamp_ms;
    const double last_ts = omega.back().timestamp_ms;
    const double imu_rate = (omega.size() > 1) ? (omega.size() - 1) / ((last_ts - first_ts) / 1000.0) : 0.0;

    double ra = (a.range_b > a.range_a) ? a.range_a : first_ts;
    double rb = (a.range_b > a.range_a) ? a.range_b : last_ts;

    std::cout << std::fixed << std::setprecision(4);
    std::cerr << "Loaded " << quats.size() << " quaternions, span [" << first_ts << ", " << last_ts
              << "] ms, IMU rate ~" << std::setprecision(2) << imu_rate << " Hz\n" << std::setprecision(4);

    if (a.mode == "omega") {
        std::cout << "timestamp_ms,wx_deg_s,wy_deg_s,wz_deg_s\n";
        for (const auto& s : omega) {
            if (s.timestamp_ms < ra || s.timestamp_ms > rb) continue;
            std::cout << s.timestamp_ms << "," << s.gyro[0] << "," << s.gyro[1] << "," << s.gyro[2] << "\n";
        }
        return 0;
    }

    if (a.mode == "compare") {
        const std::vector<double> fts = frameTimestamps(ra, rb, a.fps);
        std::vector<TimeImu> video = resampleAngularVelocity(omega, fts);
        const double ma = maxAngle(video);
        std::cerr << "Video signal: " << video.size() << " frames @ " << a.fps
                  << " fps, max |omega| = " << ma << " deg/s\n";
        if (ma < 3.0) { std::cerr << "warning: motion below 3 deg/s gate; result unreliable\n"; }
        OffsetResult r = findOffset(video, omega, a.initial, a.search, a.fps, imu_rate, a.lpf);
        if (r.found) {
            std::cout << "recovered_offset_ms=" << r.offset_ms << " cost=" << r.cost
                      << " matched=" << r.matched << "/" << video.size() << "\n";
            std::cout << "(true offset is 0; recovered value is the algorithm's bias floor)\n";
        } else {
            std::cout << "no acceptable offset found\n";
            return 1;
        }
        return 0;
    }

    if (a.mode == "selftest") {
        const std::vector<double> fts = frameTimestamps(ra, rb, a.fps);
        std::mt19937 rng(12345);                 // fixed seed -> reproducible
        std::normal_distribution<double> noise(0.0, a.noise);

        std::cout << "injected_ms,recovered_ms,error_ms,cost,matched,frames\n";
        std::vector<double> errs;
        std::size_t ok = 0;
        for (double inj : a.inject) {
            // Sample the true motion as the camera would have seen it, delayed by `inj`.
            std::vector<double> shifted;
            shifted.reserve(fts.size());
            for (double t : fts) shifted.push_back(t - inj);
            std::vector<TimeImu> video = resampleAngularVelocity(omega, shifted);
            for (std::size_t k = 0; k < video.size(); ++k) {
                video[k].timestamp_ms = fts[k]; // video timestamps stay on the video clock
                if (a.noise > 0.0) {
                    video[k].gyro[0] += noise(rng);
                    video[k].gyro[1] += noise(rng);
                    video[k].gyro[2] += noise(rng);
                }
            }
            OffsetResult r = findOffset(video, omega, a.initial, a.search, a.fps, imu_rate, a.lpf);
            if (r.found) {
                const double err = r.offset_ms - inj;
                errs.push_back(err);
                ++ok;
                std::cout << inj << "," << r.offset_ms << "," << err << "," << r.cost
                          << "," << r.matched << "," << video.size() << "\n";
            } else {
                std::cout << inj << ",NA,NA,NA,0," << video.size() << "\n";
            }
        }

        if (!errs.empty()) {
            double sum = 0.0, sq = 0.0, maxabs = 0.0;
            for (double e : errs) { sum += e; sq += e * e; maxabs = std::max(maxabs, std::abs(e)); }
            const double mean = sum / errs.size();
            const double rms = std::sqrt(sq / errs.size());
            std::cerr << "\n=== Timestamp-sync precision (" << ok << "/" << a.inject.size()
                      << " recovered) ===\n"
                      << "mean error  = " << mean << " ms\n"
                      << "RMS error   = " << rms << " ms\n"
                      << "max |error| = " << maxabs << " ms\n";
        } else {
            std::cerr << "no offsets recovered\n";
            return 1;
        }
        return 0;
    }

    std::cerr << "unknown mode: " << a.mode << "\n";
    usage();
    return 2;
}

File: py_autosync/gyroflow_autosync.py
#!/usr/bin/env python3
"""gyroflow_autosync (Python) — UI-free autosync-time tool + timestamp-sync accuracy evaluator.

Python sibling of cpp_core's ``gyroflow_autosync``. Loads DJI fused-quaternion telemetry,
derives an angular-velocity signal from the quaternions, and runs Gyroflow's time-offset finder
to report the timestamp delay (ms) between two motion signals.

Modes:
  sync      Find the timestamp offset between TWO real angular-velocity signals (e.g. an IMU
            GCSV log + a camera-motion CSV). This is the production path on real footage.
  selftest  Inject known time offsets into a video-rate copy of the signal and measure recovery
            accuracy -> timestamp-sync precision (ground-truthed).
  compare   Find the offset between the full-rate signal and a video-rate resample (true offset 0)
            -> baseline bias floor.
  omega     Dump the quaternion-derived angular velocity as CSV.

Examples:
  ./gyroflow_autosync.py sync --gyro imu.gcsv --video camera_motion.csv --interp-parabolic
  ./gyroflow_autosync.py selftest --quat ../data/dji_quaternions_full.csv --fps 30 --noise 1.5
"""

from __future__ import annotations

import argparse
import sys

import numpy as np

import autosync_time as at


def _frame_timestamps(a: float, b: float, fps: float) -> np.ndarray:
    if fps <= 0.0 or b <= a:
        return np.array([])
    dt = 1000.0 / fps
    n = int(np.floor((b - a) / dt + 1e-6)) + 1
    return a + np.arange(n) * dt


def _run_sync(args, interp: bool, parabolic: bool) -> int:
    """Sync two real angular-velocity signals (production path on real footage)."""
    if not args.gyro or not args.video:
        print("sync mode needs both --gyro and --video", file=sys.stderr)
        return 2

    gyro = at.load_motion(args.gyro, units=args.units, orientation=args.gyro_orientation)
    video = at.load_motion(args.video, units=args.units, orientation=args.video_orientation)
    gyro_rate = at.estimate_sample_rate_hz(gyro)
    video_rate = at.estimate_sample_rate_hz(video)

    g_ma, v_ma = at.max_angle(gyro), at.max_angle(video)
    print(f"gyro : {len(gyro)} samples, ~{gyro_rate:.1f} Hz, span "
          f"[{gyro.t[0]:.1f},{gyro.t[-1]:.1f}] ms, max|omega| {g_ma:.2f} deg/s", file=sys.stderr)
    print(f"video: {len(video)} samples, ~{video_rate:.1f} Hz, span "
          f"[{video.t[0]:.1f},{video.t[-1]:.1f}] ms, max|omega| {v_ma:.2f} deg/s", file=sys.stderr)
    if min(g_ma, v_ma) < 3.0:
        print("warning: motion below the 3 deg/s gate; offset may be unreliable", file=sys.stderr)
    if gyro_rate <= 0.0 or video_rate <= 0.0:
        print("error: could not estimate a sample rate (need >=2 samples per signal)", file=sys.stderr)
        return 1

    r = at.find_offset(video, gyro, args.initial, args.search, video_rate, gyro_rate, args.lpf,
                       interp=interp, parabolic=parabolic)
    if not r.found:
        print("no acceptable offset found (try a larger --search, check motion overlap/units)",
              file=sys.stderr)
        return 1
    mode = "interp+parabolic" if parabolic else ("interp" if interp else "nearest")
    print(f"offset_ms={r.offset_ms:.4f} cost={r.cost:.4f} matched={r.matched}/{len(video)} mode={mode}")
    print("(gyro samples align to video timestamps shifted by -offset_ms)", file=sys.stderr)
    return 0


def main(argv=None) -> int:
    p = argparse.ArgumentParser(description="Python autosync-time / timestamp-sync evaluator")
    p.add_argument("mode", choices=["sync", "selftest", "compare", "omega"])
    p.add_argument("--quat", help="DJI quaternion CSV (selftest/compare/omega modes)")
    p.add_argument("--gyro", help="sync mode: IMU log (GCSV or angular-velocity CSV)")
    p.add_argument("--video", help="sync mode: camera-motion angular-velocity CSV")
    p.add_argument("--units", choices=["deg", "rad"], default="deg",
                   help="sync mode: angular-velocity units when not inferable from column names")
    p.add_argument("--gyro-orientation", default=None,
                   help="sync mode: 3-char axis remap for the IMU signal, e.g. xzY")
    p.add_argument("--video-orientation", default=None,
                   help="sync mode: 3-char axis remap for the camera signal")
    p.add_argument("--fps", type=float, default=30.0, help="video frame rate for the video-side signal")
    p.add_argument("--search", type=float, default=200.0, help="offset search size (ms)")
    p.add_argument("--initial", type=float, default=0.0, help="initial/rough offset (ms)")
    p.add_argument("--lpf", type=float, default=20.0, help="low-pass cutoff (Hz)")
    p.add_argument("--swap-xy", action="store_true", help="swap x/y when deriving angular velocity")
    p.add_argument("--interp", action="store_true",
                   help="interpolated IMU lookup (removes the nearest-sample quantization bias; "
                        "breaks bit-for-bit Rust/C++ parity)")
    p.add_argument("--interp-parabolic", action="store_true",
                   help="interp + parabolic sub-grid vertex of the cost curve (us-level; implies --interp)")
    p.add_argument("--inject", type=str, default="-30,-15,-5,0,5,15,30",
                   help="selftest: comma-separated injected offsets (ms)")
    p.add_argument("--noise", type=float, default=0.0, help="selftest: Gaussian noise stddev (deg/s)")
    p.add_argument("--range", type=str, default="", help="analyse only timestamps in A,B (ms)")
    p.add_argument("--seed", type=int, default=12345, help="selftest noise RNG seed")
    args = p.parse_args(argv)

    parabolic = args.interp_parabolic
    interp = args.interp or parabolic

    if args.mode == "sync":
        return _run_sync(args, interp, parabolic)

    if not args.quat:
        p.error(f"--quat is required for the {args.mode} mode")

    t_ms, quats = at.load_quaternions(args.quat)
    omega = at.quaternions_to_angular_velocity(t_ms, quats, swap_xy=args.swap_xy, degrees=True)

    first_ts, last_ts = float(omega.t[0]), float(omega.t[-1])
    imu_rate = (len(omega) - 1) / ((last_ts - first_ts) / 1000.0) if len(omega) > 1 else 0.0

    ra, rb = first_ts, last_ts
    if args.range:
        parts = [float(x) for x in args.range.split(",") if x]
        if len(parts) == 2 and parts[1] > parts[0]:
            ra, rb = parts

    print(f"Loaded {len(quats)} quaternions, span [{first_ts:.3f}, {last_ts:.1f}] ms, "
          f"IMU rate ~{imu_rate:.2f} Hz", file=sys.stderr)

    if args.mode == "omega":
        print("timestamp_ms,wx_deg_s,wy_deg_s,wz_deg_s")
        for i in range(len(omega)):
            t = omega.t[i]
            if t < ra or t > rb:
                continue
            w = omega.w[i]
            print(f"{t:.4f},{w[0]:.4f},{w[1]:.4f},{w[2]:.4f}")
        return 0

    if args.mode == "compare":
        fts = _frame_timestamps(ra, rb, args.fps)
        video = at.resample_angular_velocity(omega, fts)
        ma = at.max_angle(video)
        print(f"Video signal: {len(video)} frames @ {args.fps} fps, max |omega| = {ma:.4f} deg/s",
              file=sys.stderr)
        if ma < 3.0:
            print("warning: motion below 3 deg/s gate; result unreliable", file=sys.stderr)
        r = at.find_offset(video, omega, args.initial, args.search, args.fps, imu_rate, args.lpf,
                           interp=interp, parabolic=parabolic)
        if r.found:
            print(f"recovered_offset_ms={r.offset_ms:.4f} cost={r.cost:.4f} "
                  f"matched={r.matched}/{len(video)}")
            print("(true offset is 0; recovered value is the algorithm's bias floor)")
            return 0
        print("no acceptable offset found")
        return 1

    # selftest
    fts = _frame_timestamps(ra, rb, args.fps)
    injects = [float(x) for x in args.inject.split(",") if x]
    rng = np.random.default_rng(args.seed)

    print("injected_ms,recovered_ms,error_ms,cost,matched,frames")
    errs = []
    ok = 0
    for inj in injects:
        video = at.resample_angular_velocity(omega, fts - inj)  # camera sees motion delayed by inj
        video.t = fts.copy()                                    # video timestamps on the video clock
        if args.noise > 0.0:
            video.w = video.w + rng.normal(0.0, args.noise, size=video.w.shape)
        r = at.find_offset(video, omega, args.initial, args.search, args.fps, imu_rate, args.lpf,
                           interp=interp, parabolic=parabolic)
        if r.found:
            err = r.offset_ms - inj
            errs.append(err)
            ok += 1
            print(f"{inj:.4f},{r.offset_ms:.4f},{err:.4f},{r.cost:.4f},{r.matched},{len(video)}")
        else:
            print(f"{inj:.4f},NA,NA,NA,0,{len(video)}")

    if errs:
        errs = np.array(errs)
        print(f"\n=== Timestamp-sync precision ({ok}/{len(injects)} recovered) ===\n"
              f"mean error  = {errs.mean():.4f} ms\n"
              f"RMS error   = {np.sqrt((errs ** 2).mean()):.4f} ms\n"
              f"max |error| = {np.abs(errs).max():.4f} ms", file=sys.stderr)
        return 0
    print("no offsets recovered", file=sys.stderr)
    return 1


if __name__ == "__main__":
    raise SystemExit(main())
